### DISCUSSION TITLE
Severe mouse stutter/input lag in osu!lazer client (M1 Mac)

### DISCUSSION THREAD

**Discussion by Penwinged** [ppy/osu#17603] [Q&A] [answered]
EDIT: To anyone else having a similar issue in the Mac osu!lazer client, I have found a solution. If you increase your track speed/cursor speed in your System Preferences and manually increase your DPI on your mouse or in osu through the sensitivity slider, it should fix any latency/input issues.

### Type

Performance

### Bug description

Problem details:
There is a noticeable issue with input latency/inconsistency for mouse input, with mouse movements being inconsistent in speed or accuracy and sometimes the cursor will decide not to move. This issue does not occur when playing other games or when using my mouse/trackpad outside of the osu client. I have tested this with a variety of different display options, frame rates and sensitivities, as well as changing the DPI of my mouse, but the problem is consistent.

The more notable examples of this in the video linked below are at timestamp 0:56-0:58 as well as 1:12-1:14, with mouse strokes either not being there entirely or seemingly having to catch up with the game at times.

Also unrelated but also a problem, the application will often decide to crash when changing between different display modes (fullscreen, windowed, borderless) which is less than ideal

My Device is a Macbook Air 2020 M1

I like playing the circle clicker game and only have a mac so if this could be looked at I would love that ty <3

### Screenshots or videos

https://youtu.be/G9CqVt5GZ3I

### Version

2022.327.0-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/8397144/database.log)
[network.log](https://github.com/ppy/osu/files/8397145/network.log)
[performance.log](https://github.com/ppy/osu/files/8397146/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8397149/runtime.log)

**Comment by frenzibyte** [comment]
The FPS view is cropped in the video, please record the full game window. Also please record with the frame graph displayed (Ctrl+F11 once more). Also please avoid using "Unlimited" rate where possible, that always tends to cause issues.

> Also unrelated but also a problem, the application will often decide to crash when changing between different display modes (fullscreen, windowed, borderless) which is less than ideal

Very likely to be https://github.com/libsdl-org/SDL/issues/5203. Can you reproduce by resizing the game window as well?

**Reply by Penwinged** [reply]
Will do so now, thanks for the response. Also I have played on multiple different frame rate modes, and the issue persists.

**Reply by frenzibyte** [reply]
Interesting, ~~the fact that the input thread is shown to run at 240-250fps is completely wrong to me~~ (turns out it was just running in single-threaded mode), so I'm curious to know what exactly is going on in the frame graphs.

**Reply by Penwinged** [reply]
https://youtu.be/q9PruBjX_FA
Full Screen Video, turned off unlimited framerate. I forgot to record with the frame graph so let me know if that is required. Thanks!

**Reply by peppy** [reply]
Frame graphs are required.

**Reply by integerrr** [reply]
Why are you running the game in singlethreaded mode?

**Reply by Coppertine** [reply]
Because Mac's are special and do singlethreaded.

**Reply by bdach** [reply]
@coppertine i'm not exactly well versed in macs but i'm pretty sure none of what you said makes any amount of sense

**Reply by Coppertine** [reply]
apologies, only judged by the comment above and thread below. won't do again

**Comment by peppy** [comment, marked as answer]
This is macOS. It isn't made for gaming. Something on your system is likely interfering with the single threaded OS event loop. It could be anything, or everything.

It's up to you to trial and error to figure which app is causing the issue. But know that even the system menu (if you have it set to auto hide) and the dock can cause these issues. It's nothing we can fix.

If your mouse is set to high polling rate, turn it down as this will reduce the overhead (1000hz is generally unusable on macOS).

The OS isn't made for gaming, sorry.

**Reply by Penwinged** [reply]
This is tragic, but thank you for the detailed response. I will attempt to find a work around. Keep up the good work!

**Reply by peppy** [reply]
Please do provide the frame graphs as requested above. I can confirm or potentially give you further advice.

**Reply by Penwinged** [reply]
@peppy I played around with my mac settings and have found a solution to the issues I've been facing. It's based on the track speed/cursor speed in System Preferences. I turned both down to the lowest possible setting and manually increased my mouse's DPI and any issues with mouse input/latency have all but gone. Not sure about the specifics as to why this is the case but thought I'd mention it. I have also edited the main post to include this information in case it helps anyone else. I am now happily playing 5* maps to my heart's content.

**Reply by peppy** [reply]
Did you happen to reduce the polling rate of your mouse in the process? That's the only thing which could have any effect.

**Reply by Penwinged** [reply]
I'm not quite sure what you're asking sorry 😅 I just manually adjusted the cursor speed of my mouse through System Preferences

**Reply by Penwinged** [reply]
Am able to run the game with default settings no problem

**Reply by integerrr** [reply]
Sorry for beating a dead horse here but I still don't believe the M1 is so incapable that the game needs to be run singlethreaded. Does running the game on multithreaded mode make it unplayable?

**Reply by peppy** [reply]
I don't know why you are posting here. The user has already resolved the issue. Threaded-ness should have no effect.

**Reply by Blackflexxy** [reply]
maybe stupid question, but does osu use this api for mouse? https://developer.apple.com/documentation/gamecontroller/gcmousemoved

**Reply by bdach** [reply]
We're using whatever [SDL](https://github.com/libsdl-org/SDL) is using. Which doesn't seem to be that but I can't say I checked very thoroughly.

**Reply by Ycat137** [reply]
I found that turning off mouse acceleration fixed the problem for me
